### PR TITLE
Proposed changes to #1229 [binary_file_test-tk2]

### PR DESCRIPTION
**Pull request by tzanio**
Proposed final adjustments to the new Travis checks in #1229.

See the comments inside the code.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ sudo: false
 stages:
   - checks
   - tests
+  - optional
```
Comment (tzanio) [inline .travis.yml]: Move the branch history check as a third stage, see https://github.com/mfem/mfem/pull/1229#discussion_r368207023.

I don't think it is easy to distinguish potential legitimate changes from potential bad commits.

File: .travis.yml [hunk 2]
```diff
@@ -73,6 +59,25 @@ jobs:
         - cd tests/scripts
         - ./runtest gitignore
 
+    # ========================
+    #  Optional Checks/Tests
+    # ========================
+    # - branch-history
+
+    - stage: optional
+      name: "branch-history"
```
Comment (tzanio) [inline .travis.yml]: `branch-history` seems more accurate than `git-history` for what this script does

File: tests/scripts/branch-history [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+if [ ${num_commits} -gt ${branch_max_commits} ]
+then
+    printf '\033[35m'"This branch contains unusually many commits: " 1>&2
+    printf "${num_commits} commits not in master"'\033[0m\n' 1>&2
+fi
+
+# Check all the commits in the current branch (i.e. the commits that will be
+# merged in master, as shown in a GitHub PR):
+while read -r rev; do
+    echo "Processing commit ${rev}..."
```
Comment (tzanio) [inline tests/scripts/branch-history]: No need to do `git checkout`

File: tests/scripts/branch-history [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+   Commit-level checks:
+   - Check if the commit changes an unusually large number of files
+   - Check if the commit changes an unusually large number of lines
+
+   File-level checks:
+   - Check if the file is binary
+   - Check if the file is unusually large
+   - Check if an unusually large number of lines have been added to the file
+   - Check if an unusually large number of lines have been deleted from the file
+
```
Comment (tzanio) [inline tests/scripts/branch-history]: Updated documentation

File: tests/scripts/branch-history [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    # Check if the commit changes an unusually large number of files
+    files_changed="$(git show --pretty="" --shortstat ${rev} | awk '{print $1}')"
+    if [ ${files_changed} -gt ${commit_max_files_changed} ]
+    then
+        printf '\033[36m'"Commit ${shortrev} changes unusually many files: " 1>&2
+        printf "${files_changed} files changed"'\033[0m\n' 1>&2
+    fi
+
+    # Check if the commit changes an unusually large number of lines
+    lines_changed="$(git show --pretty="" --shortstat ${rev} | awk '{print $4+$6}')"
```
Comment (tzanio) [inline tests/scripts/branch-history]: This seems to work just fine even if there are no insertions or deletions...

File: tests/scripts/branch-history [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+# Maximum number of acceptable lines deleted from any file changed in any commit in the branch
+file_max_lines_deleted=1000
+
+current_branch="$(git symbolic-ref --short HEAD)"
+
+# Check if the branch contains unusually large number of commits
+num_commits="$(git rev-list --count master..${current_branch})"
+if [ ${num_commits} -gt ${branch_max_commits} ]
+then
+    printf '\033[35m'"This branch contains unusually many commits: " 1>&2
```
Comment (adrienbernede) [inline tests/scripts/branch-history]: "more commits than usual"

File: tests/scripts/branch-history (hunk 3 above)
Comment (adrienbernede) [reply tests/scripts/branch-history]: We need to checkout out to check if the file is empty (per file checks) and to get the file size, because the file may have been deleted since.

File: tests/scripts/branch-history (hunk 5 above)
Comment (adrienbernede) [reply tests/scripts/branch-history]: So be it.

File: tests/scripts/branch-history [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+while read -r obj; do
+    obj_fname="$(echo ${obj} | awk '{print $2}')"
+    obj_size="$(echo ${obj} | awk '{print int($1/1024)}')"
+    if [ ${obj_size} -gt ${file_max_size} ]
+    then
+        printf '\033[35m'"This branch adds/modifies a large file: " 1>&2
+        printf "${obj_fname} (size ${obj_size}K)"'\033[0m\n' 1>&2
+    fi
+done < <(git rev-list --objects --no-merges ${commit_range} | git cat-file --batch-check='%(objecttype) %(objectsize) %(rest)' | sed -n 's/^blob //p' | awk '{if (max[$2]<$1) max[$2]=$1} END {for (f in max) print max[f], f}' | sort --numeric-sort --key=1)
+
```
Comment (tzanio) [inline tests/scripts/branch-history]: Added Veselin's test on "branch" level. It could be useful...

File: tests/scripts/branch-history [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+   To be useful, the script should be run in a clean repository that doesn't
+   have any uncommitted changes and is not in a "detached HEAD" state.
+
+   Example usage: $0 ../..
+                  $0 ../../ -commit 8d50f48
+                  $0 ../../ -range eb2017b..5dd5511
+                  $0 ../../ -range master..d5f6f7e
+                  $0 ../../ -commit d5f6f7e -fast > /dev/null
+
```
Comment (tzanio) [inline tests/scripts/branch-history]: Tests for bad commits and ranges

File: tests/scripts/branch-history (hunk 5 above)
Comment (tzanio) [reply tests/scripts/branch-history]: See https://travis-ci.org/tzanio/mfem/jobs/639714032

File: tests/scripts/branch-history (hunk 3 above)
Comment (tzanio) [reply tests/scripts/branch-history]: Good point. I addressed the tests that use `du` and `file`, but you are right that we need to wrap the `-e` too

Comment (tzanio) [reply tests/scripts/branch-history]: Fixed in 51eb265d6964

File: tests/scripts/branch-history (hunk 4 above)
Comment (theownerof) [reply tests/scripts/branch-history]: What the hell is going on here

File: tests/scripts/branch-history (hunk 7 above)
Comment (theownerof) [reply tests/scripts/branch-history]: Why is this on my phone

File: tests/scripts/branch-history [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+mfem_dir=`pwd`
+shift
+
+current_branch="$(git symbolic-ref --short HEAD)"
+commit_range="master..${current_branch}"
+if [ "$1" = "-commit" ]; then
+    shift
+    # check just the specified commit
+    commit="$1"
+    commit_range="`git rev-parse ${commit}^`..${commit}"
```
[the remark this answers is not in the document]
Comment (tzanio) [reply tests/scripts/branch-history]: Feel free to change it if that is better :)

File: tests/scripts/branch-history (hunk 6 above)
Comment (tzanio) [reply tests/scripts/branch-history]: I think the current version sounds stronger... I suggest we keep it.

File: tests/scripts/branch-history (hunk 7 above)
Comment (tzanio) [reply tests/scripts/branch-history]: Yeah, that line is kinda fun :)

Comment (adrienbernede) [reply tests/scripts/branch-history]: Agreed, and I learned some awk (dynamic creation of a map, nice).
For the sake of readability, I reformatted it, but changed nothing else.

**Note: this portion on code is sensitive to apparently insignificant changes**
- Quotes ["] in `done <<< "${git_objects}" are mandatory.
- Usage of a while loop on read is also necessary to loop on lines (a for loop would loop on "words".

File: tests/scripts/branch-history (hunk 9 above)
[the remark this answers is not in the document]
Comment (adrienbernede) [reply tests/scripts/branch-history]: Couldn’t demonstrate what I thought was a corner case. So I left it as is.

File: tests/scripts/branch-history (hunk 8 above)
Comment (adrienbernede) [reply tests/scripts/branch-history]: Great!
I haven’t tested is but it means we could use is in interactive rebase:
`> git rebase -i master --exec tests/scripts/branch_history ./ -commit HEAD -fast`
Or something similar.

File: tests/scripts/branch-history (hunk 4 above)
Comment (adrienbernede) [reply tests/scripts/branch-history]: Slightly changed usage to separate the 3 cases.
